cirq/google/calibration/workflow.py

```diff
@@ -337,19 +351,150 @@ def run_characterizations(
+
+    Returns:
+        Calibrated circuit together with its calibration metadata in CircuitCalibration object. The
+        calibrated circuit has single-qubit Z gates added which compensates for the true gates
+        imperfections.
+    """
+    if len(circuit_calibration.circuit) != len(circuit_calibration.moment_allocations):
+        raise ValueError('Moment allocations does not match circuit length')
+
+    default_phases = PhasedFSimCharacterization(zeta=0.0, chi=0.0, gamma=0.0)
+
+    compensated = Circuit()
+    compensated_allocations = []
+    for moment, characterization_index in zip(
+        circuit_calibration.circuit, circuit_calibration.moment_allocations
+    ):
+        parameters = None
+        if characterization_index is not None:
+            parameters = characterizations[characterization_index]
+
+        decompositions = []
+        other = []
+        new_moment_mapping = None
+        for op in moment:
+            if not isinstance(op, GateOperation):
+                raise IncompatibleMomentError(
+                    'Moment contains operation different than GateOperation'
+                )
+
+            if isinstance(op.gate, (MeasurementGate, SingleQubitGate)):
+                other.append(op)
+            else:
+                a, b = op.qubits
+
+                translated_gate = gates_translator(op.gate)
+                if translated_gate is None:
+                    raise IncompatibleMomentError(
+                        f'Moment {moment} contains unsupported non-single qubit operation {op}'
+                    )
+
+                if parameters is None:
+                    raise ValueError(f'Missing characterization data for moment {moment}')
+
+                pair_parameters = parameters.get_parameters(a, b)
+                if pair_parameters is None:
+                    raise ValueError(
+                        f'Missing characterization data for pair {(a, b)} in {parameters}'
+                    )
+                pair_parameters = pair_parameters.merge_with(default_phases)
+
+                corrected = PhaseCorrectedFSimOperations(
+                    (a, b), translated_gate, pair_parameters, characterization_index
+                )
+                decompositions.append(corrected.operations)
+
+                if new_moment_mapping is None:
+                    new_moment_mapping = corrected.moment_allocations
+                else:
+                    assert (
+                        new_moment_mapping == corrected.moment_allocations
+                    ), f'Inconsistent decompositions with a moment {moment}'
+
+        if other and decompositions:
+            raise IncompatibleMomentError(f'Moment {moment} contains mixed operations')
+        elif other:
+            compensated += Moment(other)
+            compensated_allocations.append(characterization_index)
+        elif decompositions:
+            for operations in zip_longest(*decompositions, fillvalue=()):
```

Comment:
why are you doing this zip(*x)? This "transposes" a nested list, no? I'm confused why it's needed here.

Reply:
The assumption is that the circuit provided here is a generic user circuit. This transposition allows me to split all the decompositions moment by moment and add those moments one by one into the circuit.